src/iocore/net/BIO_fastopen.cc

```diff
@@ -21,12 +21,72 @@
  *  limitations under the License.
  */
 
+#include <openssl/crypto.h>
+#include <openssl/opensslv.h>
+
 #include "P_Net.h"
 #include "iocore/eventsystem/SocketManager.h"
 #include "tscore/ink_assert.h"
 
 #include "BIO_fastopen.h"
 
+namespace
+{
+
+#if !defined(OPENSSL_IS_BORINGSSL)
+
+class ExData
+{
+public:
+  static int
+  idx()
+  {
+    static int idx_ = []() -> int {
+      int i = CRYPTO_get_ex_new_index(CRYPTO_EX_INDEX_BIO, 0, nullptr, _new, _dup, _free);
+      ink_release_assert(i >= 0);
+      return i;
+    }();
+    return idx_;
+  }
+
+private:
+  static void
+  _new(void * /* parent */, void * /* ptr */, CRYPTO_EX_DATA *ad, int idx_, long /* argl */, void * /* argp */)
+  {
+    ink_release_assert(CRYPTO_set_ex_data(ad, idx_, nullptr) == 1);
+  }
+
+  static void
+  _free(void * /* parent */, void * /* ptr */, CRYPTO_EX_DATA *ad, int idx_, long /* argl */, void * /* argp */)
+  {
+    ink_release_assert(CRYPTO_set_ex_data(ad, idx_, nullptr) == 1);
+  }
+
+#if OPENSSL_VERSION_MAJOR >= 3
+  using _Type_from_d = void **;
+#else
+  using _Type_from_d = void *;
+#endif
+
+  static int
+  _dup(CRYPTO_EX_DATA * /* to */, const CRYPTO_EX_DATA * /* from */, _Type_from_d /* from_d */, int /* idx */, long /* argl */,
+       void * /* argp */)
+  {
+    ink_assert(false);
+    return 0;
+  }
+};
+
+#undef BIO_set_data
+#define BIO_set_data(BIO, PTR) ink_assert(BIO_set_ex_data((BIO), ExData::idx(), (PTR)) == 1)
```

Comment:
These definitions only apply within this file (BIO_fastopen.cc).  Boringssl does not provide BIO_set/get_ex_data.  It seemed like the best approach to avoid further #ifdefs for OPENSSL_IS_BORINGSSL in this file.

Reply:
https://commondatastorage.googleapis.com/chromium-boringssl-docs/bio.h.html#ex_data-functions

Reply:
It's only been supported for about 8 months:  https://github.com/google/boringssl/commit/2139aba2e3e28cd1cdefbd9b48e2c31a75441203 .   This commit does not seem to be in the version of BoringSSL we are using in the Rocky CI job.  I guess we could add a CMake check for it.

Reply:
Bummer. I update BoringSSL once in a while so I didn't think it's unavailable.

I'd rather update the BoringSSL requirement if that's the reason. Oldest commit we currently require is https://boringssl.googlesource.com/boringssl/+/a1843d660b47116207877614af53defa767be46a
but I don't think there's a big reason to stick with the commit.
https://lists.apache.org/thread/d6ymfjhpmjogb5stnz92hwhhps909809


What do you think? Would it make the code simple?